docs/architecture/adr-012-multisigs.md

```diff
@@ -0,0 +1,135 @@
+
+The pubkey for a `k of n` multisig should be encoded as `var_uint_encode(enum_type) || var_uint_encode(k) || var_uint_encode(n) || MarshalKeys(keys)`, where `keys` is a list of `n` pubkeys.
+
+Marshal Keys should length prefix each pubkey\*, but other than that concatenate them all. In pseudocode:
+
+```
+MarshalKeys(keys [n]crypto.Pubkey) = var_uint_encode(len(keys[0])) || keys[0].Marshal || var_uint_encode(len(keys[1])) || keys[1].Marshal || ... || var_uint_encode(len(keys[n - 1])) || keys[n - 1].Marshal 
+```
+
+The weighted multi-sig pubkey will be encoded similarly.
+Let `k` be the threshold, `weight_i` is the weighting of the `i`th key, and `keys` is a list of `n` pubkeys. 
+Then the pubkey should be encoded as `var_uint_encode(enum_type) || var_uint_encode(k) || var_uint_encode(n) || var_uint_encode(weight_0) || ... || var_uint_encode(weight_{n-1}) ||  MarshalKeys(keys)` 
+
+Before defining how the multisignature will be encoded, first we define the bitmap encoding. 
+For `MarshalBitArray(indices []int, n int)`, creates a bitmap of size `⌈n / 8⌉` bytes. 
+A `1` in any of the leading `n` bits of the bitmap indicates whether or not that pubkey is represented in the set. 
+
+For a `k of n` multisig, there must be exactly `k` `1` bits. 
+For a weighted multisig the sum of the weights of the assoc. keys must be greater than the threshold. 
+There cannot be a single invalid signature, even if the sum of the correct ones is greater than the threshold.
+
+Both types of multisignatures (weighted and unweighted) encode the same. 
+It begins with the bitmap encoding of the indices.\*\* 
+It will be followed by a list of `k` signatures, encoded with the same length prefixing and appending strategy as specified in `MarshalKeys`.\*\*\*
+
+#### Footnotes
+
+\* We cannot have it derive length from the pubkey enum type (proposed [here]((https://github.com/tendermint/go-crypto/issues/121)), to avoid length prefixing. This is because we want to support a `k` of `n` multisig of ASM's (this achieves different functionality than a weighted multi sig). 
+ASM's don't have constant size pubkeys. 
+The same log stands for length prefixing signatures.
+
+\*\* The reason this doesn't use `log_8(n)` bytes per signer is because that heavily optimizes for the case where a very small number of signers are required.
+e.g. for `n` of size `24`, that would only be more space efficient for `k < 3`. 
+This seems less likely, and that it should not be the case optimized for.
+
+\*\*\* Important decision: 
+There is advocacy for using amino encoding of an `[][]byte`, instead of having a signature blob and parsing by reading the length prefix. (Note both are doing the same internally) 
+Protobuf requires having a "field number" on each signature (this is a design choice in protobuf encoding of slices), which means an extra byte on all signatures relative to the proposed encoding scheme, in addition to a constant overhead.
```

Comment:
Haven't read this in detail to have an informed opinion yet. Just to let you know: while in proto3 a fieldnumber is mandatory, the amino API allows to encode (length-prefixed) raw bytes, without a field number. Not sure if this good or bad (my first intuition would be that this shouldn't be doable with amino). We even make sue of this already in secret connection (no fields / structs involved):
https://github.com/tendermint/tendermint/blob/2aa2b63cadc42cca1071c36adfd2f2ce14e1aa8f/p2p/conn/secret_connection.go#L199-L213